### PR TITLE
Clean up ccommon rust bindings

### PR DESCRIPTION
**Pull request by swlynch99**
This is the first of a sequence of PRs for parts of #260. This one cleans up the ccommon bindings and modifies the cmake files to better support rust. Specifically,
- Only generate bindings for ccommon, not all of libc
- Add bindings for ccommon's `buf` type to ccommon_rs
- Add a top level Cargo.toml to group all the projects and remove the existing smaller workspaces within the repo
- Change testing to be done through just calling `cargo test` in the root directory instead of special casing the rust tests in the build script

*Note:* The tests under `ccommon_rs` appear to never have been run under CI and one of the ones that create temporary files appear to break. I've disabled this one for now and will address it in a follow up commit once I understand more about how the logging code in `ccommon_rs` works.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,11 +1,11 @@
-##-- cdb Cargo.toml file --##
-# note: do not change the mark above #
```
Comment (brayniac) [inline Cargo.toml]: Whitespace here is a little odd.

File: deps/ccommon/include/buffer/cc_buf.h [hunk 2]
```diff
@@ -68,7 +68,7 @@ struct buf {
     char              *wpos;    /* write marker */
     char              *end;     /* end of buffer */
     bool              free;     /* is this buf free? */
-    char              begin[1]; /* beginning of buffer */
+    char              begin[];  /* beginning of buffer */
```
Comment (brayniac) [inline deps/ccommon/include/buffer/cc_buf.h]: @thinkingfish will this cause any issues?

File: deps/ccommon/rust/cc_binding/Cargo.toml [hunk 3]
```diff
@@ -2,10 +2,14 @@
...
+edition = "2018"
 
 [build-dependencies]
-bindgen = "0.37.0"
-failure = "~0.1.1"
+bindgen = "0.51.1"
+failure = "0.1.5"
+
+[dependencies]
+libc = "0.2"
```
Comment (brayniac) [inline deps/ccommon/rust/cc_binding/Cargo.toml]: Please add the bugfix version to the libc dep.

File: deps/ccommon/rust/cc_binding/build.rs [hunk 4]
```diff
@@ -3,57 +3,64 @@ extern crate bindgen;
 extern crate failure;
 
 use std::env;
+use std::ffi::OsString;
 use std::fs;
 use std::io;
```
Comment (brayniac) [inline deps/ccommon/rust/cc_binding/build.rs]: You can combine this `use` with the one for `std::io::BufReader` eg: `use std::io::{self, BufReader}`

I agree with leaving the one including `prelude::*` on its own line.

File: deps/ccommon/rust/cc_binding/build.rs [hunk 5]
```diff
@@ -3,57 +3,64 @@ extern crate bindgen;
...
-use std::io::BufReader;
 use std::io::prelude::*;
-use std::ffi::OsString;
+use std::io::BufReader;
 use std::os::unix::fs as unix_fs;
 use std::path::Path;
```
Comment (brayniac) [inline deps/ccommon/rust/cc_binding/build.rs]: I know it was already like this... but can you change the two use declarations for `std::path` to a single use?

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 6]
```diff
@@ -0,0 +1,292 @@
+// ccommon - a cache common library.
+// Copyright (C) 2018 Twitter, Inc.
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Incorrect year. I think we can make the license header the shorter version. eg:
```
// Copyright 2019 Twitter, Inc.
// Licensed under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0
```

Defer to @thinkingfish on whether it's appropriate to use the even shorter header.

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        (**self).write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        (**self).flush()
+    }
+}
+
+
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Whitespace

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Read for OwnedBuf {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        (**self).read(buf)
+    }
+}
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Add newline between trait implementations

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        let len = self.read_size().min(out.len());
+
+        unsafe {
+            std::ptr::copy_nonoverlapping(self.buf.rpos as *const u8, out.as_mut_ptr(), len)
+        }
+        self.buf.rpos = self.buf.rpos.wrapping_add(len);
+
+        Ok(len)
+    }
+}
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Add newline between trait implementations

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        &mut self.buf as *mut _
+    }
+
+    /// The number of bytes that can still be written to the
+    /// buffer before it is full.
+    pub fn write_size(&self) -> usize {
+        assert!(self.buf.wpos as usize <= self.buf.end as usize);
+
+        self.buf.end as usize - self.buf.wpos as usize
+    }
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Add newline

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+/// a new instance (via transmute) will cause UB.
+#[repr(transparent)]
+pub struct Buf {
+    buf: buf
+}
+
+impl Buf {
+    pub unsafe fn from_raw<'a>(buf: *const buf) -> &'a Buf {
+        &*(buf as *const Buf)
+    }
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Add newline

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+
+impl Deref for OwnedBuf {
+    type Target = Buf;
+
+    fn deref(&self) -> &Buf {
+        unsafe {
+            Buf::from_raw(self.buf)
+        }
+    }
+}
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: add newline

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Resize the buffer to fit the required capacity.
+    /// 
+    /// # Panics
+    /// Panics if `cap` is greater than `u32::MAX`.
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Do we really want to take a `usize` here then? Might make more sense to take a `u32` size?

File: deps/ccommon/rust/ccommon_rs/src/buf.rs [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+/// an `OwnedBuf` also supports some resizing operations.
+#[repr(transparent)]
+pub struct OwnedBuf {
+    buf: *mut buf
+}
+
+impl OwnedBuf {
+    pub unsafe fn from_raw(buf: *mut buf) -> Self {
+        Self { buf }
+    }
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/buf.rs]: Add newline

File: deps/ccommon/rust/ccommon_rs/src/lib.rs [hunk 15]
```diff
@@ -37,8 +37,39 @@ use std::result;
 pub mod bstring;
 pub mod log;
 pub mod util;
+pub mod buf;
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/lib.rs]: Please sort alphabetically

File: deps/ccommon/rust/ccommon_rs/src/log/mod.rs [hunk 16]
```diff
@@ -609,7 +609,7 @@ pub unsafe extern "C" fn log_create_handle_rs(cfgp: *mut bind::log_config_rs) ->
     ptrs::null_check(cfgp)                                // make sure our input is good
         .map_err(|e| e.into())                            // error type bookkeeping
         .and_then(|c|LogConfig::from_raw(c))              // convert the *mut into a rust struct
-        .and_then(log_setup_safe)                         // register our logger
+        .and_then(log_setup)                         // register our logger
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/log/mod.rs]: re-align the comments

File: deps/ccommon/rust/ccommon_rs/src/util.rs [hunk 17]
```diff
@@ -41,3 +43,82 @@ pub unsafe extern "C" fn cc_util_rm_rf_rs(path: *const c_char) -> i32 {
...
+
+        // Panic in cases that obviously won't work
+        assert_eq!(size_of_val(self) % size_of::<metric>(), 0);
+
+        let count = size_of_val(self) / size_of::<metric>();
+        slice::from_raw_parts(
+            self as *const _ as *const metric,
+            count
+        )
+    }
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/util.rs]: Add newline

File: deps/ccommon/rust/ccommon_rs/src/util.rs [hunk 18]
```diff
@@ -41,3 +43,82 @@ pub unsafe extern "C" fn cc_util_rm_rf_rs(path: *const c_char) -> i32 {
...
+
+        // Panic in cases that obviously won't work
+        assert_eq!(size_of_val(self) % size_of::<option>(), 0);
+
+        let count = size_of_val(self) / size_of::<option>();
+        slice::from_raw_parts(
+            self as *const _ as *const option,
+            count
+        )
+    }
```
Comment (brayniac) [inline deps/ccommon/rust/ccommon_rs/src/util.rs]: Add newline

File: deps/ccommon/rust/ccommon_rs/src/buf.rs (hunk 13 above)
Comment (swlynch99) [reply deps/ccommon/rust/ccommon_rs/src/buf.rs]: Generally anything in rust that represents a size is done as a `usize`. If we use a `u32` here then I suspect most call sites will just do `some_usize as u32` which would silently give wrong results when `some_usize > u32::MAX`. That being said, we can change this if you think that `u32` is a better choice.

File: deps/ccommon/rust/ccommon_rs/src/buf.rs (hunk 6 above)
Comment (swlynch99) [reply deps/ccommon/rust/ccommon_rs/src/buf.rs]: I've fixed the date for this. Will replace the license header if we decide to go with the shorter one.

File: deps/ccommon/include/buffer/cc_buf.h (hunk 2 above)
Comment (swlynch99) [reply deps/ccommon/include/buffer/cc_buf.h]: One note here: using `begin[1]` is technically undefined behaviour. Idk if any compilers exploit it but using `begin[]` is explicitly supported by the C standard.

A second note: I didn't find any code using `sizeof(buf)` anywhere within either pelikan or ccommon.

Comment (thinkingfish) [reply deps/ccommon/include/buffer/cc_buf.h]: I'm almost sure the incorrect declaration is a copy of how `struct item` is defined in the slab module in twemcache (https://github.com/twitter/twemcache/blob/master/src/mc_items.h#L134), which came from memcache

File: deps/ccommon/rust/cc_binding/Cargo.toml [hunk 19]
```diff
@@ -2,10 +2,14 @@
 name = "cc_binding"
 version = "0.1.0"
 authors = ["Jonathan Simms <jsimms@twitter.com>"]
```
Comment (thinkingfish) [inline deps/ccommon/rust/cc_binding/Cargo.toml]: Feel free to add your name to it, @swlynch99

File: deps/ccommon/include/buffer/cc_buf.h (hunk 2 above)
Comment (brayniac) [reply deps/ccommon/include/buffer/cc_buf.h]: Okay with the change here then

File: deps/ccommon/rust/ccommon_rs/src/buf.rs (hunk 13 above)
Comment (brayniac) [reply deps/ccommon/rust/ccommon_rs/src/buf.rs]: Per discussion. OK to leave as usize, fix the assert to be `<=`

Comment (thinkingfish) [reply deps/ccommon/rust/ccommon_rs/src/buf.rs]: For the scope of this project (i.e. we don't expect ccommon to be widely used by itself) it seems fine that we use `u32` to simplify the signature.

**Pull request merged**